# AAE722 Homework 1
Bruce Cai

In [93]:
import numpy as np
import pandas as pd
from matplotlib.pyplot import subplots
import statsmodels.api as sm
from ISLP import load_data
from ISLP.models import (ModelSpec as MS,
                         summarize)
from ISLP import confusion_table
from ISLP.models import contrast
from sklearn.discriminant_analysis import \
     (LinearDiscriminantAnalysis as LDA,
      QuadraticDiscriminantAnalysis as QDA)
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

## 1 Question1

In [94]:
data = pd.read_csv("D://desktop//AAE722//raw_data//Advertising.csv")
data

Unnamed: 0     TV  Radio  Newspaper  Sales
0             1  230.1   37.8       69.2   22.1
1             2   44.5   39.3       45.1   10.4
2             3   17.2   45.9       69.3    9.3
3             4  151.5   41.3       58.5   18.5
4             5  180.8   10.8       58.4   12.9
..          ...    ...    ...        ...    ...
195         196   38.2    3.7       13.8    7.6
196         197   94.2    4.9        8.1    9.7
197         198  177.0    9.3        6.4   12.8
198         199  283.6   42.0       66.2   25.5
199         200  232.1    8.6        8.7   13.4

[200 rows x 5 columns]

### 1.1 Replicating the Regression Table


In [95]:
X = MS(['TV', 'Radio','Newspaper']).fit_transform(data)
y = data["Sales"]
model1 = sm.OLS(y, X)
results1 = model1.fit()
print(summarize(results1))

             coef  std err       t  P>|t|
intercept  2.9389    0.312   9.422   0.00
TV         0.0458    0.001  32.809   0.00
Radio      0.1885    0.009  21.893   0.00
Newspaper -0.0010    0.006  -0.177   0.86


### 1.2 Hypotheses for the p-values
For "TV", the null hypothesis (H<sub>0</sub>) is: &beta;<sub>TV</sub>=0, which suggests that TV advertising has no effect on sales. A close to 0 p-value would reject this null hypothesis, indicating that TV advertising has a statistically significant effect on sales.

For "Radio", the null hypothesis (H<sub>0</sub>) is: &beta;<sub>Radio</sub>=0, which suggests that Radio advertising has no effect on sales. Similar to TV, a close to 0 p-value would reject this null hypothesis, indicating that Radio advertising has a statistically significant effect on sales.


For "Newspaper", the null hypothesis (H<sub>0</sub>) is: &beta;<sub>Newspaper</sub>=0, which suggests that Newspaper advertising has no effect on sales. The p-value=0.86 is very high. Thus we fail to reject the null hypothesis, meaning there is no statistically significant relationship between newspaper advertising and sales based on this model.

### 1.3 Interpreting the Results
Both TV and radio advertising have significant positive effects on sales, with radio having a stronger estimated effect than TV. The coefficients for TV and Radio are 0.0458 and 0.1885. Meas that for every 1 unit increase in TV advertising spending, sales are expected to increase by 0.0458 units. And for every 1 unit increase in Radio advertising spending, sales are expected to increase by 0.1885 units.
 
There is no statistically significant relationship between newspaper and sales, suggesting that newspaper advertising may not be an effective channel to increasing sales in this case.

# 2 Question2
KNN Classifier: KNN Classifier is a classification method that categorizes the most similar K samples based on the similarity of the independent variables among the samples by the given parameter K.

KNN Regression: KNN regression is a regression method based on KNN classification. After classify the sample using KNN, the target is predicted by calculating the mean of the K samples that are most similar to the other variables of the predicted target.

# 3 Question3

(a) I would expect the cubic regression has lower training RSS than linear. Because the cubic regression model is based on a linear model (when &beta;<sub>2</sub>=&beta;<sub>3</sub>=0 it's a linear model), there is more flexibility to fit the training set when using cubic regression.

(b) But there is anthoer situation when it's "test" data set. Cubic regression will capture the noise of the training set due to overfitting, making his test set perform less well than linear. So the RSS for cubic regression will probably larger.

(c) I would expect the cubic regression has lower training RSS than linear. Because the cubic model is more flexible due to it's higher order terms (X<sup>2</sup>, X<sup>3</sup>), and can approximate a wider variety of relationships between X and Y. 

(d) I would also expect the cubic regression has lower testing RSS than linear since its flexibility and the true founction it's non-linear. But in some special case it's hard to tell which model will have lower RSS, for example when the true founction is close to linear but not linear.


# 4 Question4

In [96]:
rng = np.random.default_rng(1)
x = rng.normal(size=100)
y = 2 * x + rng.normal(size=100)

(a)

In [97]:
model1 = sm.OLS(y, x).fit()
print(summarize(model1))



      coef  std err       t  P>|t|
x1  1.9762    0.117  16.898    0.0


In this model we do not include the intercept term, which is the same as in the true founction. The estimate &beta; is 1.97 close to the true &beta;=2. And the p-value is 0.0 means we rejects the null hypothesis, the relationship between x and y is statistically significant.

(b)

In [98]:
model2 = sm.OLS(x, y).fit()

print(summarize(model2))


      coef  std err       t  P>|t|
x1  0.3757    0.022  16.898    0.0


The coefficient estimate is 0.3757, smaller than I expected around 0.5, the p-value is still close to 0, indicating a significant result.

(c) I would have expected the two coefficients to be the inverse of each other roughly 2 and 0.5, but the second coefficient is only 0.3757. I'm guessing it's because the theoretical coefficient of 0.5 when regressing x onto y is affected by the  error term "rng.normal(size=100)", while there is no INTERCEPT term to adjust the model not pass the zero point. The t-statistic is same.

(d) 
$$
\hat{\beta} = \left( \sum_{i=1}^{n} x_i y_i \right) \Bigg/ \left( \sum_{i=1}^{n} x_i^2 \right)
$$
since $(y_i - x_i \hat{\beta})^2 = ϵ^2$, and the sum of it is SSE
$$
\text{SE}(\hat{\beta}) = \sqrt{ \frac{\sum_{i=1}^{n} (y_i - x_i \hat{\beta})^2}{(n - 1) \sum_{i'=1}^{n} x_{i'}^2} }
$$
$$
\hat{\beta}/\text{SE}(\hat{\beta}) = \frac{(\sqrt{n-1})\sum_{i=1}^{n} x_i y_i}{\sqrt{SSE \sum_{i=1}^{n} x_i^2}}
$$


In [99]:
beta_hat = sum(x*y)/sum(x**2)
se_beta = (sum((y-x*beta_hat)**2)/(99*sum(x**2)))**0.5
t_1 = beta_hat/se_beta
t_2 = ((99**0.5)*sum(x*y))/((sum(x**2)*sum(y**2)-(sum(x*y))**2)**0.5)
print(t_1,t_2)
print("t_1 = t_2:",round(t_1,4)==round(t_2,4))

16.8984170630351 16.89841706303509
t_1 = t_2: True


(e) Since the t-statistic can be calculated by:
$$
t = \frac{(\sqrt{n-1}) \sum_{i=1}^{n} x_i y_i}{\sqrt{\left( \sum_{i=1}^{n} x_i^2 \right) \left( \sum_{i=1}^{n} y_i^2 \right) - \left( \sum_{i=1}^{n} x_i y_i \right)^2}}
$$
there is no difference when we swap x and y in this formula. So the t-statistic should remain the same.

(f)

In [104]:
df = pd.DataFrame({'x': x, 'y': y})
X = MS(['x']).fit_transform(df)
model1 = sm.OLS(y, X)
results1 = model1.fit()
print(summarize(results1))
Y = MS(['y']).fit_transform(df)
model2 = sm.OLS(x, Y)
results2 = model2.fit()
print(summarize(results2))


             coef  std err       t  P>|t|
intercept -0.0760    0.101  -0.756  0.451
x          1.9686    0.118  16.734  0.000
             coef  std err       t  P>|t|
intercept  0.0095    0.044   0.216  0.829
y          0.3763    0.022  16.734  0.000
